### PR TITLE
[op-conductor] sequencer health monitor

### PR DESCRIPTION
**Pull request by 0x00101010**
**Description**

Implement sequencer health monitor.

**Tests**

Added tests for testing all failure scenarios for the monitor.

### CODE REVIEW

File: op-node/p2p/peer_scores_test.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package p2p
+package p2p_test
```
Comment (0x00101010) [inline op-node/p2p/peer_scores_test.go:1]: Need to move the test into `p2p_test` package to avoid cyclic imports caused by the new mock (API). This is the common practice anyway.

Other changes are just auto formatting by goimports

File: op-node/p2p/rpc_api.go [hunk 2]
```diff
@@ -20,8 +20,8 @@ type PeerInfo struct {
 	ENR             string   `json:"ENR"`       // might not always be known, e.g. if the peer connected us instead of us discovering them
 	Addresses       []string `json:"addresses"` // multi-addresses. may be mix of LAN / docker / external IPs. All of them are communicated.
 	Protocols       []string `json:"protocols"` // negotiated protocols list
-	//GossipScore float64
-	//PeerScore float64
+	// GossipScore float64
+	// PeerScore float64
```
Comment (refcell) [inline op-node/p2p/rpc_api.go:24]: FWIW Generally better practice to split smaller style changes out into a separate pr so as to not bloat the pr diff, but no big deal this time around

File: op-conductor/health/monitor.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		hm.log.Error("health monitor failed to get peer stats", "err", err)
+		return false
+	}
+	if uint64(stats.Connected) < hm.minPeerCount {
+		hm.log.Error("peer count is below minimum", "connected", stats.Connected, "minPeerCount", hm.minPeerCount)
+		return false
+	}
+
+	return true
+}
```
Comment (refcell) [inline op-conductor/health/monitor.go]: Unless I'm misunderstanding, the `lastSeenUnsafeBlock` and `lastSeenSafeBlock` are reset each loop iteration, so I think we can make them local instead of `SequencerHealthMonitor` state variables. Something like:

```suggestion
func (hm *SequencerHealthMonitor) healthCheck() bool {
	ctx := context.Background()
	status, err := hm.node.SyncStatus(ctx)
	if err != nil {
		hm.log.Error("health monitor failed to get sync status", "err", err)
		return false
	}
	lastSeenUnsafeBlock := status.UnsafeL2
	lastSeenSafeBlock := status.SafeL2

	now := uint64(time.Now().Unix())
	// allow at most one block drift for unsafe head
	if now - lastSeenUnsafeBlock.Time > hm.interval + hm.rollupCfg.BlockTime {
		hm.log.Error("unsafe head is not progressing", "lastSeenUnsafeBlock", lastSeenUnsafeBlock)
		return false
	}

	if now - lastSeenSafeBlock.Time > hm.safeInterval {
		hm.log.Error("safe head is not progressing", "safe_head_time", lastSeenSafeBlock.Time, "now", now)
		return false
	}

	stats, err := hm.p2p.PeerStats(ctx)
	if err != nil {
		hm.log.Error("health monitor failed to get peer stats", "err", err)
		return false
	}
	if uint64(stats.Connected) < hm.minPeerCount {
		hm.log.Error("peer count is below minimum", "connected", stats.Connected, "minPeerCount", hm.minPeerCount)
		return false
	}

	return true
}
```

File: op-conductor/health/monitor.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	done chan struct{}
+	wg   sync.WaitGroup
+
+	rollupCfg           *rollup.Config
+	safeInterval        uint64
+	minPeerCount        uint64
+	interval            uint64
+	healthUpdateCh      chan bool
+	lastSeenUnsafeBlock eth.L2BlockRef
+	lastSeenSafeBlock   eth.L2BlockRef
```
Comment (refcell) [inline op-conductor/health/monitor.go]: Can remove, see below
```suggestion
```

**Review by refcell — COMMENTED** [review]
Overall looks great!

File: op-node/p2p/rpc_api.go (hunk 2 above)
Comment (0x00101010) [reply op-node/p2p/rpc_api.go:24]: Totally, I thought about not doing the style changes as well, but kept it due to the small diff size.

Will keep this in mind for future PRs if there's large formatting changes

File: op-conductor/health/monitor.go (hunk 3 above)
Comment (0x00101010) [reply op-conductor/health/monitor.go]: Good call, removed.

File: op-conductor/health/monitor.go (hunk 4 above)
Comment (0x00101010) [reply op-conductor/health/monitor.go]: Removed

File: op-node/p2p/rpc_api.go (hunk 2 above)
Comment (refcell) [reply op-node/p2p/rpc_api.go:24]: Awesome thank you 👍

**Review by refcell — APPROVED** [review]
Looks good to me! Just curious, is there a ticket / reason for implementing this?

Also, would ask @tynes to take a peak as well 😄

**Comment by 0x00101010** [comment]
> Looks good to me! Just curious, is there a ticket / reason for implementing this?
> 
> Also, would ask @tynes to take a peak as well 😄

We're working on `op-conductor` to introduce sequencer HA features, currently Matt Slipper, Zach Howard, and Joshua Gutow are collaborating with us (Base) for this.

I've created an issue to track this: https://github.com/ethereum-optimism/protocol-quest/issues/44

**Comment by tynes** [comment]
Thoughts on adding the HTTP endpoint [here](https://github.com/ethereum-optimism/optimism/tree/develop/op-exporter#health-endpoint-health) so that we can delete `op-exporter`?

File: op-conductor/health/monitor.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+			hm.healthUpdateCh <- hm.healthCheck()
+		}
+	}
+}
+
+// healthCheck checks the health of the sequencer by 3 criteria:
+// 1. unsafe head is progressing per block time
+// 2. safe head is progressing every configured batch submission interval
+// 3. peer count is above the configured minimum
+func (hm *SequencerHealthMonitor) healthCheck() bool {
```
Comment (tynes) [inline op-conductor/health/monitor.go:107]: I notice this pings the `op-node` RPC API, would it also be useful to ping the EL RPC API? This would ensure that both services are actually running

Comment (0x00101010) [reply op-conductor/health/monitor.go:107]: We definitely could. 
My current understanding is that unsafeHead will not proceed if EL is not healthy, can somebody help confirm? If not, we can add an extra step here to ping EL RPC to make sure it is alive.

**Comment by 0x00101010** [comment]
> Thoughts on adding the HTTP endpoint [here](https://github.com/ethereum-optimism/optimism/tree/develop/op-exporter#health-endpoint-health) so that we can delete `op-exporter`?

Two points to discuss:
1. we'll need to provide a client library for identifying the current leader, seems like something we might want to do anyway (regardless of the health check thing)
2. As you mentioned before, this couples sequencer health check with op-conductor, if somebody else chose not to use HA setup, then op-exporter is still needed. We can definitely add a health endpoint on op-conductor, but that does not seem to allow us to delete op-exporter.
  a. One idea is introduce stand-alone mode for op-conductor, which it does not run consensus, just run health checks and expose health endpoints (this could allow us to run op-conductor in single sequencer mode and delete op-exporter)

**Review by zhwrd — APPROVED** [review]

File: op-conductor/health/monitor.go (hunk 5 above)
Comment (mslipper) [reply op-conductor/health/monitor.go:107]: Correct, unsafehead cannot proceed unless the EL is healthy.

**Review by mslipper — APPROVED** [review]

**Comment by mslipper** [comment]
I think we should deal with op-exporter separately. We may be able to just delete it outright; I don't think we're running it anywhere in our infrastructure and there have been better ways of monitoring sequencer health for quite some time.
